src/modules/tokens.ts

```diff
@@ -1,34 +1,111 @@
 import { Address, BigInt } from "@graphprotocol/graph-ts"
 import { ERC20 } from "../../generated/GPV2Settlement/ERC20"
 import { Token } from "../../generated/schema"
-
-const DEFAULT_DECIMALS = 18
+import { ZERO_BD, ZERO_BI } from "../utils/constants"
+import { ERC20SymbolBytes } from '../../generated/Factory/ERC20SymbolBytes'
+import { ERC20NameBytes } from '../../generated/Factory/ERC20NameBytes'
+import { StaticTokenDefinition } from '../utils/staticTokenDefinition'
+import { isNullEthValue } from '../utils'
 
 export namespace tokens {
 
-    export function getOrCreateToken(tokenAddress: Address, timestamp: BigInt): Token {
-        let tokenId = tokenAddress.toHexString()
-        let token = Token.load(tokenId)
-      
-        if (!token) {
-          token = new Token(tokenId)
-          token.address = tokenAddress
-          token.firstTradeTimestamp = timestamp
-      
-          let erc20Token = ERC20.bind(tokenAddress)
-          let tokenDecimals = erc20Token.try_decimals()
-          let tokenName = erc20Token.try_name()
-          let tokenSymbol = erc20Token.try_symbol()
-          token.decimals = !tokenDecimals.reverted
-              ? tokenDecimals.value
-              : DEFAULT_DECIMALS
-          token.name = !tokenName.reverted ? tokenName.value : ""
-          token.symbol = !tokenSymbol.reverted ? tokenSymbol.value : ""
-      
+  export function getOrCreateToken(tokenAddress: Address, timestamp: BigInt): Token {
+    let tokenId = tokenAddress.toHexString()
+    let token = Token.load(tokenId)
+
+    if (!token) {
+      token = new Token(tokenId)
+      token.address = tokenAddress
+
+      token.decimals = fetchTokenDecimals(tokenAddress) 
+      token.name = fetchTokenName(tokenAddress)
+      token.symbol = fetchTokenSymbol(tokenAddress)
+      token.derivedETH = ZERO_BD
+      token.derivedUSD = ZERO_BD
+      token.allowedPools = []
+
+    }
+    // adding timestamp only if it's a trade from Cow
+    // trades from Uniswap will call this function with ZERO_BI value
```

Comment:
then, i would recommend to make it more explicit in the param names.

nit: Can the `timestamp` be optional instead of assuming a `ZERO_BI` value?

Also, can the name be something like  `cowTradeTimestamp`?

Reply:
By looking to the signature, is not very clear to me i can call this function without the timestamp until i read the implementation